### ISSUE TITLE
Block Templates: Editing Products by Category creates new Taxonomy templates

### ISSUE THREAD

**Issue by tjcafferkey** [woocommerce/woocommerce-gutenberg-products-block#9275]
## Describe the bug

After editing and saving the Products by Category, it creates a _new_ template titled "Taxonomy Product Cat 2" which you cannot delete or remove. It can end up creating multiple of these unremovable templates as seen in the below screenshot.

It also specifies this template has been created from the WooCommerce plugin.

<img width="735" alt="Screenshot 2023-04-28 at 10 40 03" src="https://user-images.githubusercontent.com/8639742/235113629-1e68cbbd-f6f7-4383-b2bc-ad5e86835562.png">

## To reproduce

1. Go to Appearance > Editor
2. Click on Templates > Products by Category to edit this template
3. Add a block and save. Check the "Manage all templates" list view to see if a new template has been created. If not, then repeat this step.
4. See additional template

## Expected behavior

It should not create a new template

## Screenshots

https://user-images.githubusercontent.com/8639742/235113163-47552395-18f3-4606-ae78-7cd5f4ca201e.mp4

### WordPress (please complete the following information):

-   WordPress version: 6.2
-   WooCommerce version: 7.6.1
-   WooCommerce Blocks version: 10.1.0

## Additional context

I am not sure yet if this is a Gutenberg issue or a WooCommerce issue.

**Comment by kmanijak** [comment]
Early findings:

1. Issue occurs even with WC Blocks disabled
2. Issue occurs also for Product by Attributes and Products by Tag
3. What makes those attempts of updating the template different from the others is the response from WP REST API returns:
```{
  "id": "woocommerce/woocommerce//taxonomy-product_cat",
  "theme": "woocommerce/woocommerce",
  "content": { ... },
  "slug": "taxonomy-product_cat",
  "source": "custom",
  "origin": null,
  "type": "wp_template",
  "description": "Displays products filtered by a category.",
  "title": {
    "raw": "Products by Category",
    "rendered": "Products by Category"
  },
  "status": "publish",
  "wp_id": 110,
  "has_theme_file": false,
  "is_custom": true,
  "author": 1,
  "_links": { ... }
}
```

and either `origin` is `null` (as in the example above) or `source` is `null`. If both are set correctly so `origin: 'plugin', source: 'custom'` then it’s all good.

4. I found out that the new template is not created until going back to the templates list by clicking "Manage All Templates". Checked by verifying the content of `wp_posts` table in DB. There's no new entry until getting back to see all of the templates.
5. The template changes are saved correctly within WP REST API. However, I wasn't able to debug `update_item` method [here](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L318-L377). For some reason the debugger wouldn't stop there, even though it stopped correctly on the previously called callback `update_item_permissions_check` [here](https://github.com/WordPress/wordpress-develop/blob/4ad3cad2e0d9369d96f04837babe6270a1423487/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L306).
6. While going to the templates list, `get_block_templates_from_db` from `BlockTemplateUtils.php` returns already a list with new incorrect template [here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/Utils/BlockTemplateUtils.php#L687).

Requires further investigation.

**Comment by thealexandrelara** [comment]
I did some initial investigation on this, here are some findings:

1. The issue is occurring even when using WooCommerce 7.8.0-dev version (I thought it could be related to the WooCommerce version that you're currently using);
2. I had to do the steps twice to get the error, so I compared both requests (the first one, that didn't make the error occur, and the second one that was right before the template being duplicated). Both of the requests have the same response header and body, the only difference was the order of the query params for the `Referer` request header (despite the params and their values are equal). Basically the first time the `postId` param came first than the `postType` and the second time, the `postId` came after the `postType`. The only way that I can think that this could possible cause an error is if the code is using the params by their position instead of their keys;
3. By inspecting the request to `wp-json/wp/v2/templates` when the "Taxonomy Product Cat 2" appeared  I noticed that the first time that the template was duplicated, the `content.raw` was empty and the `content.blockVersion` was set to 0 for this template. When running the steps again and the "Taxonomy Product Cat 3" appeared, the `content.raw` was duplicated from the original template and the `content.blockVersion` was set to 1 for this new template.

**Comment by thealexandrelara** [comment]
@tjcafferkey continued the investigation and has found that besides the changes on the `source` and `origin` properties, each response return a different `wp_id` as well.

Considering that, I was investigating the behavior for the `wp_id`, and here is what I found:

1. Let's assume the original template  (Products by Category) has an ID of `642`.
2. When we edit, and save the template, the API endpoint `/wp-json/wp/v2/templates/woocommerce/woocommerce//taxonomy-product_cat` is called. Upon checking the response, we notice that the ID returned is no longer `642`, but a different number, for instance, `650`.
3. Clicking on "Manage All Templates" triggers a call to `/wp-json/wp/v2/templates`, which returns a list of all available templates. By checking the response, we can verify that the "Products by Category" template now has the same ID as the one returned in the previous step, `650`. Furthermore, no other template will appear in the list, suggesting that the ID was simply replaced by a new one.
4. If we once again click on the "Products by Category" template, edit it, and save it, we'll notice that the ID returned on the server's response will be the initial ID, `642`
5. After returning to the "All Templates" page and checking the API response (`/wp-json/wp/v2/templates`), we'll notice that the original "Products by Category" template will has an ID of `642`. In addition, a new template called "Taxonomy Product Cat 2" has been created with an ID of `650`.

In summary, the first time the Products by Category template is saved, the id is replaced by a new one (for example: from `642` to `650`). The second time that we save the template, the id is replaced by the previous one (`642`), while the new id generated (`650`) is set to a newly created template ("Taxonomy Product Cat 2").

Not sure if this could be a cache-related issue; it will require further investigation.

**Comment by tjcafferkey** [comment]
**TLDR;** we unset the `source` property [here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTemplatesController.php#L127) in the `BlockTemplatesController.php` which is causing WordPress to create a new database entry for every save _after_ the first save.

----

We filter the return value of [this function](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L319) in our Controller [here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTemplatesController.php#L66). This results in the `$template` object not having a `source` property and because of this you can see in the endpoint that [it creates a new database entry as a result of this](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L342-L349).

Removing [this line](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTemplatesController.php#L127) fixes the issue but this seems intentionally put here so I am wondering why this is the case, looks like it was introduced as part of https://github.com/woocommerce/woocommerce-blocks/pull/7975.

Going to investigate further the initial fix and see if I can open a PR without introducing any regressions.
